### ISSUE TITLE
How to load partition data into dataloader in cross process setting

### ISSUE THREAD

**Issue by xuannn1998** [SMILELab-FL/FedLab#234]
Hi, 

I'm trying fedavg with **cross process** setting now.
How to load data into dataloader (for each client) after I partition trainset into `shards_part`?

<img width="486" alt="截圖 2022-08-16 00 03 02" src="https://user-images.githubusercontent.com/44284861/184671589-cbaa6ffb-23e7-423c-9e0a-780c3bdeac37.png">

**Comment by AgentDS** [comment]
Since partitioner has partitioned dataset for each client into a `dict` {cid: [data_idx1, data_idx2,...]}, just load the corresponding data using this `dict` after loading the whole dataset. For detailed example, please check [_get_dataloader](https://github.com/SMILELab-FL/FedLab-benchmarks/blob/6df5ead7ed40094d579caaff0a6822ddb3a854ca/fedlab_benchmarks/feddyn/client.py#L45)

**Comment by xuannn1998** [comment]
Can I use SubsetSampler instead?
```
trainset = torchvision.datasets.CIFAR10(
        root=data_dir, train=True, download=True, transform=transforms.ToTensor())

num_shards = 200
num_clients = 10
seed = 2021

partition = CIFAR10Partitioner(
    trainset.targets, 
    num_clients,
    balance=None, 
    partition="shards",
    num_shards=num_shards,
    seed=seed)

trainloader = torch.utils.data.DataLoader(
    trainset,
    sampler=SubsetSampler(indices=partition[args.rank-1], shuffle=True),
    batch_size=args.batch_size)
```

**Comment by AgentDS** [comment]
Yes, you can. While it can be kind of slow since each client only gets hundreds of samples from more than O(10^4) samples. I recommend you to use `Subset` (check [here](https://github.com/SMILELab-FL/FedLab-benchmarks/blob/6df5ead7ed40094d579caaff0a6822ddb3a854ca/fedlab_benchmarks/feddyn/client.py#L38)).

**Comment by AgentDS** [comment]
try learning rate decay?

**Comment by AgentDS** [comment]
Have you tried our demo code? Please check [this](https://github.com/SMILELab-FL/FedLab-benchmarks/blob/6df5ead7ed40094d579caaff0a6822ddb3a854ca/fedlab_benchmarks/feddyn/client.py#L187). I haven't tried it when client_num=10, but it achieves acceptable performance when client_num=100. Try learning rate and other strategies in this demo.

**Comment by xuannn1998** [comment]
I tried this demo code (**cross process setting**) in [this folder](https://github.com/SMILELab-FL/FedLab/tree/f038b70c021c26caa179c8e8a9f46c7cb735b93f/examples/cross-process-mnist).
I change the dataset and model to cifar-10 and cnn-cifar10.

**Comment by AgentDS** [comment]
The demo you tried may have problem in hyperparameter setting. We didn't tune hyperparameter in this demo, just showing the simple demo. Please try [this](https://github.com/SMILELab-FL/FedLab-benchmarks/blob/6df5ead7ed40094d579caaff0a6822ddb3a854ca/fedlab_benchmarks/feddyn/client.py#L187), which has partial result report.

**Comment by xuannn1998** [comment]
Thanks.

I found that **label-skew distribution** in the FMNIST partitioner will give each user different number of samples except for #C=1 (major_class_num=1). Is it possible to have each client get the same number of samples?
```
FMNISTPartitioner(
    trainset.targets, 
    num_clients=num_clients,
    partition="noniid-#label", 
    major_classes_num=2,
    seed=seed)
```
<img width="524" alt="截圖 2022-08-20 15 37 47" src="https://user-images.githubusercontent.com/44284861/185734562-e402c303-2dcc-4d78-a972-5229b0c89669.png">

**Comment by AgentDS** [comment]
It depends on the data partition you want to have. We call such partition as 'balanced' if it supports same number partition. Some partitions support 'balanced', while some not. You can also try to make your own partition scheme depending on your experimental setting, constating the same sample number for each client. Functions in [this](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/utils/dataset/functional.py) may help if you want to create your own partition.
